compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -0,0 +1,181 @@
+            // 3. All arg0 refer to the same value.
+            // 4. All arg1 are IrConst<*>.
+
+            if (condition.symbol != codegen.classCodegen.context.irBuiltIns.eqeqSymbol)
+                return false
+
+            // TODO: The check relies on the implementation that subject is always lowered to a local variable.
+            val candidate = condition.getValueArgument(0) as? IrGetValue ?: return false
+            subject = subject ?: candidate
+            if (candidate.symbol != subject!!.symbol)
+                return false
+            if (!candidate.type.isInt())
+                return false
+
+            val case = condition.getValueArgument(1) as? IrConst<*> ?: return false
+            if (case.kind != IrConstKind.Int)
+                return false
+
+            cases.add(Pair(case.value as Int, thenLabel))
+
+            return true
+        } else if (condition is IrWhen && condition.origin == IrStatementOrigin.WHEN_COMMA) {
+            // WHEN_COMMA should always be a Boolean.
+            assert(condition.type.isBoolean())
+
+            // Match the following structure:
+            //
+            // when() {
+            //   cond_1 -> true
+            //   cond_2 -> true
+            //   ...
+            //   else -> cond_N
+            // }
+            //
+            // Namely, the structure which returns true if any one of the condition is true.
+            for (branch in condition.branches) {
+                if (branch is IrElseBranch) {
+                    if (!branch.condition.isTrueConst())
+                        return false
+                    if (!matchConditions(branch.result, thenLabel))
+                        return false
+                } else {
+                    if (!branch.result.isTrueConst())
+                        return false
+                    if (!matchConditions(branch.condition, thenLabel))
+                        return false
+                }
+            }
+
+            return true
+        }
+
+        return false
+    }
+
+    private fun gen(expression: IrElement, data: BlockInfo): StackValue = codegen.gen(expression, data)
+
+    private fun coerceNotToUnit(fromType: Type, fromKotlinType: KotlinType?, toKotlinType: KotlinType): StackValue =
+        codegen.coerceNotToUnit(fromType, fromKotlinType, toKotlinType)
+
+    fun gen(): StackValue {
+        cases.sortBy { it.first }
+
+        // Emit the temporary variable for subject.
+        gen(subject!!, data)
+
+        // Emit either tableswitch or lookupswitch, depending on the code size.
+        //
+        // lookupswitch is 2X as large as tableswitch with the same entries. However, lookupswitch is sparse while tableswitch must
+        // enumerate all the entries in the range.
+        val caseMin = cases.first().first
+        val caseMax = cases.last().first
+        val rangeLength = caseMax - caseMin + 1L
+        if (rangeLength > 2L * cases.size) {
```

Comment:
@punzki Old codegen also checks range inerval:
```
if (rangeLength > 2L * labelsNumber || rangeLength > Int.MAX_VALUE) 
```
Please move this code to old codegen companion and reuse here

Reply:
I think you mean @ting-yuan :-)

Reply:
@punzki Sorry!

Reply:
Done.